### PR TITLE
Relate bytes to IO specification

### PR DESCRIPTION
**Pull request by mlimbeck**
Includes helper functions to relate concrete bytes with the abstract representation of the network packet in the IO spec.

### CODE REVIEW

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:31]: let's drop these lines if they are not necessary

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:23]: The spacing in the file seems to be inconsistent. Please address it everywhere. In particular, do not have two or more consecutive empty lines, and do not have two or more consecutive spaces together.

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:387]:
```suggestion
pure func computePrevSegLen(currHF int, seg1Len int, seg2Len int, seg3Len int) (res int){
```

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:399]: these lines are quite long and hard to read. please, try to extract repeated subexpressions to new let bindings and make sure that all let bindings are aligned

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:500]: can you introduce an intermediate pure func to avoid having to write all these unfoldings explicitly?

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:588]: are all these unfoldings necessary? None of the lets use the underlying memory of the `raw` slice in the RHS.

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:608]: these should have a body as well. Is the definition for these functions unclear in the spec/Isabelle model?

**Review by jcp19 — CHANGES_REQUESTED** [review]

File: router/dataplane_spec.gobra
Comment (mlimbeck) [reply router/dataplane_spec.gobra]: Done.

File: router/dataplane_spec.gobra
Comment (mlimbeck) [reply router/dataplane_spec.gobra:23]: I removed all consecutive spaces except the ones in the predicate. Should I remove those as well?

File: router/dataplane_spec.gobra
Comment (mlimbeck) [reply router/dataplane_spec.gobra]: Done.

File: router/dataplane_spec.gobra
Comment (mlimbeck) [reply router/dataplane_spec.gobra]: Done.

File: router/dataplane_spec.gobra
Comment (mlimbeck) [reply router/dataplane_spec.gobra]: Done.

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:414]: I think that we should pick the trigger explicitly here (`res[k]` may be a good choice)

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:416]: can you break this signature into multiple lines to make it easier to read?

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:429]: can we pick the trigger here?

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:528]: this function has very large lines, can you please split them?

**Review by jcp19 — COMMENTED** [review]
I started taking a look at this but I haven't looked at it super in depth. I will do that later this week. I have a few requests already that have to do with readability

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:359]: `computeSeqLen` is a misleading name. I would call it sth like `lengthOfCurrSeg`. Also, I wonder if we need to attach all these preconditions. I know what you mean by them, but here they are only adding extra requirements that must be checked, even though this call can never fail regardless of the values of the outputs. The advantage is that we can get more localized error messages, if it is indeed the case that this condition must hold on all call sites

File: verification/io/other_defs.gobra
Comment (jcp19) [inline verification/io/other_defs.gobra:233]: where does the implementation for this function come from? How is it related to the IO spec in Isabelle?

Comment (jcp19) [reply verification/io/other_defs.gobra:233]: Nvm, I got it. I wonder if we should add preconditions here enforcing which constructors were used for building the msg term.

**Review by jcp19 — COMMENTED** [review]
So far, things look good. I wonder if the performance problems that you reported on our last meeting manifest in the code from this PR already

**Comment by jcp19** [comment]
@mlimbeck can you please give a status update on this PR?

**Comment by mlimbeck** [comment]
Once we have merged PR #230 I am going to look how we can use the new ghost state `DataPlaneSpec` to solve the issue with the `asid` argument in `getHopfieldsConsDir()` and `getHopfieldsNotConsDir()`

**Review by Dspil — APPROVED** [review]

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:697]: could we even use `hopFieldOffset` here?

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:732]: Please fix the spacing here and in the following functions
```suggestion
pure func getCurrINF(hdr uint32) uint8 {
```

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:758]: Don't we have anything like that in the package `slayers/path`? If so, please use those instead. Otherwise, maybe it should exist there (or rather, in the package where we parse these fields from the Header, instead of this package - in a sense, we are exposing implementation details from some package, thus it should be in that package. Also, In general, I fell like functions named `get...` are very unidiomatic in Go. I would call these `CurrHF`, `LenSeg1`, `LenSeg2`, ...

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:808]:
```suggestion
	return currHFIdx == len(hopfields) ?
		getHVFSet(hopfields[currHFIdx-1]) :
		(currHFIdx == 0 ?
				getHVFSet(hopfields[currHFIdx]) :
				(consDir ?
						getHVFSet(hopfields[currHFIdx]) :
						getHVFSet(hopfields[currHFIdx-1]))
```

**Review by jcp19 — CHANGES_REQUESTED** [review]
I have a few minor comments to make this more readable. Should be ready to merge after they are addressed

File: pkg/slayers/path/hopfield_spec.gobra
Comment (jcp19) [inline pkg/slayers/path/hopfield_spec.gobra:27]: If this is meant to be used outside of the package, then we should make it public. Even though Gobra does not enforce this now, it will likely do so in the future

File: pkg/slayers/path/hopfield_spec.gobra
Comment (mlimbeck) [reply pkg/slayers/path/hopfield_spec.gobra:47]: I will need those function to prove `DecodeFromBytes()` for hopfields. If I move it to`path/scion` I have the same cyclic path problem.

File: pkg/slayers/path/hopfield_spec.gobra
Comment (mlimbeck) [reply pkg/slayers/path/hopfield_spec.gobra:27]: In this case it is meant to be private.

File: pkg/slayers/path/scion/base_spec.gobra
Comment (jcp19) [inline pkg/slayers/path/scion/base_spec.gobra:53]: where is this used?

File: router/dataplane_spec.gobra
Comment (jcp19) [inline router/dataplane_spec.gobra:25]: do we actually need these imports in this file?

File: router/io-spec.gobra
Comment (jcp19) [inline router/io-spec.gobra:39]: here, we seem to be redefining the function (there is a copy of it on file `no-filed_spec.gobra`). If that is the case, then let's drop one of them

File: pkg/slayers/path/hopfield_spec.gobra
Comment (jcp19) [inline pkg/slayers/path/hopfield_spec.gobra:47]: are we redefining these functions in the `router` package? If so, let's drop one copy of each

File: router/io-spec.gobra
Comment (jcp19) [inline router/io-spec.gobra:181]:
```suggestion
		(left == none[seq[io.IO_as]] || right == none[seq[io.IO_as]]) ?
			none[seq[io.IO_as]] : 
			some(get(left) ++ get(right)[1:])
```

**Review by jcp19 — APPROVED** [review]
Looks ok, there are a few minor points that should be addressed before merging. After this is handled and the CI passes, we can merge it

File: pkg/slayers/path/scion/base_spec.gobra
Comment (mlimbeck) [reply pkg/slayers/path/scion/base_spec.gobra:53]: I created this function to make sure that the new ghost variable `path.MetaLen` is equal to the original variable but I am not really sure how to use this function. I though if I ensure the result, it implies that the variables are equal?

Comment (jcp19) [reply pkg/slayers/path/scion/base_spec.gobra:53]: ah I see. I would put this in file base_spec_test.gobra to make it clear that it is a test (and not a member that we use somewhere)

Comment (mlimbeck) [reply pkg/slayers/path/scion/base_spec.gobra]: Alright, function is moved. It is ready to merge.

**Pull request merged**
